### PR TITLE
Feature/path adherence feature

### PR DESCRIPTION
**Pull request by silanus23**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #5037 |
| Primary OS tested on | Ubuntu|
| Robotic platform tested on |turtlebot3|
| Does this PR contain AI generated software? |Yes and it is marked inline in the code)|

---

## Description of contribution in a few bullet points

* Added a new utility function, nav2_util::distanceFromPath, to efficiently calculate the shortest distance from a robot's pose to the global path. This utility is optimized with an iterative search index and correctly handles closed-loop paths.

* Added a new lifecycle publisher to the controller_server to broadcast the path tracking error on a dedicated topic (/tracking_error). Starts with movement. 

* Added a new, self-contained Behavior Tree condition node, IsFollowingPath, which subscribes to the error topic  (/tracking_error) and returns FAILURE if the deviation exceeds a configurable threshold. This allows for creating high-level, reactive recovery behaviors.

* Added a new DWB critic, PathHugCritic, which uses the new utility to penalize trajectories based on their average deviation from the global path, enforcing stricter path adherence.

* Added a new MPPI critic, PathHugCritic, which implements the same path-hugging logic for the MPPI controller.

## Description of documentation updates required from your changes

* DWB pathHug.scale

* MPPI PathHugCritic.path_hug_weight, PathHugCritic.path_hug_power.

* The new IsFollowingPath Behavior Tree node, along with its input ports (max_error, error_topic), should be added.

## Description of how this change was tested

* Added unit tests (mostly made by AI) and tested on turtlebot3 gazebo but had hard time to test it properly. This subject needs some help from maintainers.

---
### Update & Request for Feedback

This pull request introduces a comprehensive feature set to enforce strict path adherence, directly addressing issue #5037. By adding a distance utility, a BT condition node, and specialized critics for both DWB and MPPI, this contribution provides Nav2 with the crucial capability to constrain path deviation for high-precision tasks where simply following the general direction of the path is insufficient.

While the core functionality and unit tests are complete, further integration testing would be beneficial to fully demonstrate the impact of these changes. I found it challenging to create simulation scenarios that clearly highlight the behavioral differences with and without the new critics, especially for MPPI and the new BT node logic.

I would greatly appreciate any guidance or suggestions from the maintainers on effective ways to test and validate these kinds of nuanced behavioral improvements.

---

## Future work that may be required in bullet points

* DWB and MPPI critics can benefit from parameters. 
* Further tests needed. Especcially in edge cases. 
---

#### For Maintainers: 
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new parameters added are updated in docs.nav2.org
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists
- [ ] Should this be backported to current distributions? If so, tag with `backport-*`.

### CODE REVIEW

File: nav2_behavior_tree/nav2_tree_nodes.xml [hunk 1]
```diff
@@ -387,7 +387,11 @@
       <input_port name="error_codes_to_check">Error codes to check, user defined</input_port>
     </Condition>
 
-
+    <Condition ID="IsFollowingPath">
```
Comment (SteveMacenski) [inline nav2_behavior_tree/nav2_tree_nodes.xml:390]: Docs: need a configuration guide page for this new BT node + add to the Navigation Plugins page

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * @brief  Behavior-Tree condition that returns SUCCESS while the robot’s
+ *         lateral tracking error is within a configurable bound.
+ *
+ * Ports:
+ *   • max_error   (double)   — Allowed deviation in metres (default: 1.0)
+ *   • error_topic (string)   — Topic publishing nav2_msgs/TrackingError
+ */
+class IsFollowingPathCondition : public BT::ConditionNode
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp:38]: How about `IsWithinPathTrackingBounds`

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+  /// Tick: SUCCESS if |error| ≤ max_error_, else FAILURE.
+  BT::NodeStatus tick() override;
+
+  static BT::PortsList providedPorts()
+  {
+    return {
+      BT::InputPort<double>("max_error", 1.0,
+        "Maximum allowed tracking error"),
+      BT::InputPort<std::string>("error_topic", "/tracking_error",
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp:54]:
```suggestion
      BT::InputPort<std::string>("error_topic", "tracking_error",
```

So it works with namespacing :-)

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  void errorCallback(const nav2_msgs::msg::TrackingError::SharedPtr msg);
+
+  // ---------------------------------------------------------------------
+  rclcpp::Node::SharedPtr node_;
+  rclcpp::Subscription<nav2_msgs::msg::TrackingError>::SharedPtr sub_;
+  rclcpp::CallbackGroup::SharedPtr cb_group_;
+  rclcpp::executors::SingleThreadedExecutor exec_;
+
+  std::string topic_;
+  double      max_error_      {1.0};
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp:70]: This probably causes linting problems, please check out the CI job outputs

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        "nav2_msgs/msg/TrackingError topic")
+    };
+  }
+
+private:
+  void initSubscriber();   // one-shot init on first tick
+  void errorCallback(const nav2_msgs::msg::TrackingError::SharedPtr msg);
+
+  // ---------------------------------------------------------------------
+  rclcpp::Node::SharedPtr node_;
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp:64]: if you store the logger, you don't need the entire node

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    const BT::NodeConfiguration & conf);
+
+
+  /// Tick: SUCCESS if |error| ≤ max_error_, else FAILURE.
+  BT::NodeStatus tick() override;
+
+  static BT::PortsList providedPorts()
+  {
+    return {
+      BT::InputPort<double>("max_error", 1.0,
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp:52]: Maybe an option for left vs right error? That can be different according to some standards.

I think also making the default a bit larger would be good i.e. 2m

File: nav2_behavior_tree/plugins/condition/is_following_path_condition.cpp [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+  RCLCPP_WARN(
+    node_->get_logger(),
+    "Tracking diverged: error %.3f > max %.3f", last_error_, max_error_);
+  return BT::NodeStatus::FAILURE;
+}
+
+}  // namespace nav2_behavior_tree
+
+// ---------------------------------------------------------------------------
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/condition/is_following_path_condition.cpp:113]: Remove these comment blocks in front of hte methods to be aligned with the styling of the other BT nodes

File: nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+ *         lateral tracking error is within a configurable bound.
+ *
+ * Ports:
+ *   • max_error   (double)   — Allowed deviation in metres (default: 1.0)
+ *   • error_topic (string)   — Topic publishing nav2_msgs/TrackingError
+ */
+class IsFollowingPathCondition : public BT::ConditionNode
+{
+public:
+  IsFollowingPathCondition(
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/plugins/condition/is_following_path_condition.hpp:41]: All functions need doxygen API docs

File: nav2_behavior_tree/plugins/condition/is_following_path_condition.cpp [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+  got_msg_ = true;
+}
+
+// ---------------------------------------------------------------------------
+//  TICK
+// ---------------------------------------------------------------------------
+BT::NodeStatus IsFollowingPathCondition::tick()
+{
+  if (!BT::isStatusActive(status())) {
+    initSubscriber();
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/condition/is_following_path_condition.cpp:85]: I believe we have some conventions for what we name this initialization function used in other BT nodes

File: nav2_behavior_tree/plugins/condition/is_following_path_condition.cpp [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+// ---------------------------------------------------------------------------
+void IsFollowingPathCondition::initSubscriber()
+{
+  if (sub_ready_) {
+    return;
+  }
+
+  // ---- read ports (defaults set in header) ----
+  getInput("error_topic", topic_);
+  getInput("max_error", max_error_);
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/condition/is_following_path_condition.cpp:49]: I don't think max error needs to be found here

File: nav2_behavior_tree/plugins/condition/is_following_path_condition.cpp [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+  exec_.spin_some();
+
+  if (!got_msg_) {
+    RCLCPP_DEBUG(node_->get_logger(), "Waiting for first TrackingError message…");
+    return BT::NodeStatus::FAILURE;
+  }
+
+  // Allow runtime update of the threshold
+  getInput("max_error", max_error_);
```
Comment (SteveMacenski) [inline nav2_behavior_tree/plugins/condition/is_following_path_condition.cpp:96]: If we obtain each iteration, this doesn't need to be a class member function

File: nav2_controller/src/controller_server.cpp [hunk 12]
```diff
@@ -262,9 +268,16 @@ ControllerServer::on_activate(const rclcpp_lifecycle::State & /*state*/)
   for (it = controllers_.begin(); it != controllers_.end(); ++it) {
     it->second->activate();
   }
+
   vel_publisher_->on_activate();
   action_server_->activate();
 
+  // Activate tracking error publisher
+  if (tracking_error_pub_) {
+    tracking_error_pub_->on_activate();
```
Comment (SteveMacenski) [inline nav2_controller/src/controller_server.cpp:277]: This is never used / populated to publish an outcome

File: nav2_bt_navigator/src/bt_navigator.cpp [hunk 13]
```diff
@@ -51,6 +52,8 @@ BtNavigator::BtNavigator(rclcpp::NodeOptions options)
     this, "odom_topic", rclcpp::ParameterValue(std::string("odom")));
   declare_parameter_if_not_declared(
     this, "filter_duration", rclcpp::ParameterValue(0.3));
+  declare_parameter_if_not_declared(
+    this, "tracking_error", rclcpp::ParameterValue(std::string("tracking_error")));
```
Comment (SteveMacenski) [inline nav2_bt_navigator/src/bt_navigator.cpp:56]: This parameter is never used

File: nav2_controller/include/nav2_controller/controller_server.hpp [hunk 14]
```diff
@@ -133,6 +134,16 @@ class ControllerServer : public nav2::LifecycleNode
    * @param name Reference to the name to use for control if any valid available
    * @return bool Whether it found a valid controller to use
    */
+  void path_error();
```
Comment (SteveMacenski) [inline nav2_controller/include/nav2_controller/controller_server.hpp:137]: This is unimplemented

File: nav2_util/include/nav2_util/path_utils.hpp [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * @return  Shortest distance in metres, or +∞ if @p path is empty.
+ *
+ * @throws std::invalid_argument if @p pose.header.frame_id differs from
+ *                               @p path.header.frame_id.
+ */
+double distanceFromPath(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const nav_msgs::msg::Path & path,
+  size_t *                                closest_idx = nullptr);
```
Comment (SteveMacenski) [inline nav2_util/include/nav2_util/path_utils.hpp:52]: Odd formatting heret hat needs to be addressed

File: nav2_util/src/path_utils.cpp [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+
+namespace
+{
+// === helpers ===============================================================
+
+constexpr double kEpsilon = 1e-8;
+constexpr double kClosedPathTol = 0.01;  // [m]
+
+inline double
+squaredDistance(
```
Comment (SteveMacenski) [inline nav2_util/src/path_utils.cpp:34]: This should exist in geometry_utils.hpp

File: nav2_msgs/msg/TrackingError.msg [hunk 17]
```diff
@@ -0,0 +1,8 @@
+# TrackingError.msg
+# Real-time tracking error for Nav2 controller
+
+std_msgs/Header header
+float32 tracking_error
```
Comment (SteveMacenski) [inline nav2_msgs/msg/TrackingError.msg:5]: If the BT node supports left/right, then we should have this be signed based on left/right. As such, then we should describe the signedness meaning in the message so developers can use this message best

File: nav2_msgs/msg/TrackingError.msg [hunk 18]
```diff
@@ -0,0 +1,8 @@
+# TrackingError.msg
```
Comment (SteveMacenski) [inline nav2_msgs/msg/TrackingError.msg:1]: If we include the robot's pose, speed, distance to goal, etc then this message doesn't need to be TrackingError, but it can be a larger tracking state message. If we do that, what other information might be useful to provide?

File: nav2_util/src/path_utils.cpp [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+#include <algorithm>
+#include <cmath>
+#include <limits>
+
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "geometry_msgs/msg/point.hpp"
+#include "nav_msgs/msg/path.hpp"
+#include "rcutils/logging_macros.h"
+
+namespace
```
Comment (SteveMacenski) [inline nav2_util/src/path_utils.cpp:26]: Why are any of these methods outside of the namespace?

File: nav2_util/src/path_utils.cpp [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+ * @param closest_idx  IN/OUT latch index; may be nullptr
+ * @return             Shortest distance [m] or +inf for empty path
+ *
+ * @throws std::invalid_argument if pose/path frame_ids differ
+ */
+double
+distanceFromPath(
+  const geometry_msgs::msg::PoseStamped & pose,
+  const nav_msgs::msg::Path & path,
+  size_t *                                closest_idx /* = nullptr */)
```
Comment (SteveMacenski) [inline nav2_util/src/path_utils.cpp:98]: This should be used to start searching from, if given. It should also be updated in this method for the newest start point for future iterations.

Why is this a pointer?

**Comment by SteveMacenski** [comment]
@silanus23 I gave this a review, but I don't think this works since nothing actually publishes or computes a value. Without being too nosy, this looks like alot of AI generated code that wouldn't work. I think we should back up and slow down to make progress. 

I think we should probably focus on one thing at a time and build up:
- [ ] Create a utility function that can find the distance from the path. If given an input last index, use that to start search from and update that for future iterations & only search up to a fixed distance ahead. Have this well tested and checked that this works with looping 
- [ ] Add a new TrackingState message to the Controller Server's feedback message which contains that information and other important tracking state info (speed, time progressing, pose, path, tracking error, current path IDX, etc)
- [ ] Use this feedback in a behavior tree node to find the tracking state and succeed/fail based off of it 
- [ ] Create a behavior tree that uses it and show that this is triggered when out of bounds so we generally know it works

Then we can actively change the behavior of the system based on that.
- [ ] Pass the tracking state into the controller plugins so that they can do things with it
- [ ] MPPI/DWB should use this information in critic functions to penalize trajectories out of bounds within a given time horizon
- [ ] Graceful/RPP should probably just ignore this information, but have the option via a parameter to reject controls if out of bounds
